File: gauss/c++/gauss.cpp
#include"gauss.h"

Gauss::Gauss(){
	n=0;//количество уравнений
	cur=0;//текущая строка или столюец, взависимости от ситуации
	max=0;//номер строки с максимальным элементом в нужном нам столбце
	}
void Gauss::readkeyboard(){//чтение с клавиатуры
	cout<<"Введите размер матрицы (n):"<<endl;
	cin>>n;
	A=new float *[n+1];//массив коэффициентов
	for(int i=0;i<n+1;++i)A[i]=new float[n+1];
	Y=new float[n+1]; //вектор, то есть правая часть уравнения
	X=new float[n+1]; //вектор иксов, то есть решений
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j){
			cout<<"Введите a["<<i<<"]["<<j<<"]:"<<endl;
			cin>>A[i][j];
			}
		cout<<"Введите y"<<i<<endl;
		cin>>Y[i];
		}
	A1=new float *[n+1];//сохраняем матрицу коэффициентов для рассчёта вектора невязки
	for(int i=0;i<n+1;++i)A1[i]=new float[n+1];
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j)A1[i][j]=A[i][j];
		}
	Y1=new float[n+1];//сохраняем вектор b для векторора невязки
	for(int i=0;i<n;++i)Y1[i]=Y[i];
	F=new float[n+1];//создаём вектор невязки
	for(int i=0;i<n;++i)F[i]=0;
	}
void Gauss::readfile(){//чтение из файла
	char *fn=new char[20];//имя файла
	ifstream fin;
	char ch=0;//просто символьная переменная в которую считываем из файла,нужна,чтобы подсчитать количество уравнений
	int num=0;//количество пробелов в первой строке
	cout<<"Введите имя файла:"<<endl;
	cin>>fn;
	fin.open(fn);
	while(ch!=10){
		fin.read((char*)&ch,1);
		if(ch==32)++num;
		}
	n=num;
	cur=0;
	max=0;
	fin.seekg(0);
	X=new float [n+1];
	Y=new float [n+1];
	A=new float *[n+1];
	for(int i=0;i<n+1;++i)A[i]=new float[n+1];
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j)fin>>A[i][j];
		fin>>Y[i];
		}
	A1=new float *[n+1];
	for(int i=0;i<n+1;++i)A1[i]=new float[n+1];
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j)A1[i][j]=A[i][j];
		}
	Y1=new float[n+1];
	for(int i=0;i<n;++i)Y1[i]=Y[i];
	F=new float [n+1];
	for(int i=0;i<n;++i)F[i]=0;
	}
void Gauss::findMax(){//поиск главного элемента
	max=cur;
	for(int i=cur;i<n;++i){
		if(fabs(A[max][cur])<fabs(A[i][cur]))max=i;
		}
	}
void Gauss::swap(){//перестановка строк
	float temp;//пустой флоат,нужен для обмена значений между переменными
	findMax();
	if(cur!=max){
		for(int i=cur;i<n;++i){
			temp=A[max][i];
			A[max][i]=A[cur][i];
			A[cur][i]=temp;
			}
		temp=Y[max];
		Y[max]=Y[cur];
		Y[cur]=temp;		
		}
	temp=A[cur][cur];
	for(int i=cur;i<n;++i){
		for(int j=cur;j<n;++j){
			A[i][j]/=temp;
			}
		Y[i]/=temp;
		}
	}
void Gauss::func(){
	float temp;//пустой флоат, нужен для приведения матрицы к векхнетреугольному виду
	for(int j=1;cur+j<n;++j){
		temp=A[cur+j][cur];
		for(int i=cur;i<n;++i){
			A[cur+j][i]-=A[cur][i]*temp;
			}
		Y[cur+j]-=Y[cur]*temp;
		}
	}
void Gauss::gauss1(){//прямой ход Гаусса
	for(;cur<n-1;++cur){
		swap();
		func();
		}
	}
void Gauss::func2(){
	float Y1=Y[cur];
	if(n!=cur){
		for(int i=n-1;i>cur;--i){
			Y1-=A[cur][i]*X[i];
			}
		}
	X[cur]=Y1/A[cur][cur];
	}
void Gauss::gauss2(){//обратный ход Гаусса
	for(;cur>=0;--cur){
		func2();
		}
	}
void Gauss::showMatrix(){
	cout<<endl<<"СЛАУ:"<<endl;
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j){
			cout<<A[i][j]<<"*X"<<j+1<<"  ";
			}
		cout<<"  =  "<<Y[i]<<endl;
		}
	}
void Gauss::showResult(){
	cout<<endl<<"Result:"<<endl;
	for(int i=0;i<n;++i){
		cout<<"X"<<i+1<<" = "<<X[i]<<endl;
		}
	}
void Gauss::calcvector(){//подсчёт вектора невязки
	for(int i=0;i<n;++i){
		for(int j=0;j<n;++j){
			F[i]+=A1[i][j]*X[j];
			}
		F[i]-=Y1[i];
		}
	}
void Gauss::showVector(){
	float norma=F[0];//норма вектора невязки
	for(int i=0;i<n;++i){
		cout<<"F["<<i<<"]="<<F[i]<<endl;
		if(fabs(norma)<fabs(F[i]))norma=fabs(F[i]);
		}
	cout<<"Норма="<<norma<<endl;
	}

File: gauss/python/gauss.py
# -*- coding: utf-8 -*-
import numpy as np
import math
import os
from time import time
Max=[0]
cur=[0]
check1=0;
check2=0;
D=[0]
N=int(raw_input("Введите число уравнений:\n N="))
A=np.zeros((N,N))
Y=np.zeros(N)
X=np.zeros(N)
F=np.zeros(N)
while check1!=1 and check1!=2: 
    check1=int(raw_input("Нажмите 1, если хотите ввести уравнения в ручную!\nНажмите 2, если хотите загрузить уравнения из файла!"))
    if check1==1:
        for i in range(N):
            for j in range(N):
                A[i,j]=float(raw_input("Введите A[%d,%d]: "%(i,j)))
            Y[i]=float(raw_input("Введите Y[%d]: "%i))
    if check1==2:
        while check2==0:
            fn=raw_input("Введите имя файла:")
            if os.path.exists(fn):
                check2=1;
                data=np.loadtxt(fn,delimiter=' ',dtype=np.float)
                for i in range(N):
                    for j in range(N):
                        A[i,j]=data[i,j]
                    Y[i]=data[i,N]
            else:
                print("Файл не найден!")
print(A,Y)
Acopy=A.copy()
Ycopy=Y.copy()
def findMax(ar,M,cu,n):
    M[0]=cu[0]
    for i in range(cu[0],n):
        if math.fabs(ar[M[0],cu])<math.fabs(ar[i,cu[0]]):
            M[0]=i
def swap(ar,y,M,cu,n):
    findMax(ar,M,cu,n)
    if cur!=M:
        for i in range(cu[0],N):
            temp=ar[M,i]
            ar[M,i]=ar[cu[0],i]
            ar[cu[0],i]=temp
        temp=y[M]
        y[M]=y[cu[0]]
        y[cu[0]]=temp
    temp=ar[cu[0],cu[0]]
    for i in range(cu[0],n):
        for j in range(cu[0],n):
            ar[i,j]=ar[i,j]/temp
        y[i]=y[i]/temp
def func1(ar,y,cu,n):
    for i in range(cu[0]+1,n):
        temp=ar[i,cu[0]]
        for j in range(cu[0],n):
            ar[i,j]-=ar[cu[0],j]*temp
        y[i]-=y[cu[0]]*temp
def gauss1(ar,y,M,cu,n):
    while(cu[0]<n-1):
        swap(ar,y,M,cu,n)
        func1(ar,y,cu,n)
        cu[0]+=1
def func2(ar,y,x,cu,n):
    y1=y[cu[0]]
    if n!=cu[0]:
        for i in range(n-1,cu[0],-1):
            y1-=ar[cu[0],i]*x[i]
    x[cu[0]]=y1/ar[cu[0],cu[0]]
def gauss2(ar,y,x,cu,n):
    while(cu[0]>=0):
        func2(ar,y,x,cu,n)
        cu[0]-=1
def calcvect(ar,y,x,f,n,d):
    for i in range(n):
        for j in range(n):
            f[i]+=ar[i,j]*x[j]
        f[i]-=y[i]
    d[0]=math.fabs(f[0])
    for i in range(n):
        if d[0]<math.fabs(f[i]):
            d[0]=math.fabs(f[i])
	    print(d[0])
tic=time()
gauss1(A,Y,Max,cur,N)
toc=time()
print("Время работы прямого хода: %f" %(toc-tic))
tic=time()
gauss2(A,Y,X,cur,N)
toc=time()
print("Время работы обратного хода: %f" %(toc-tic))
calcvect(Acopy,Ycopy,X,F,N,D)
print("Решение системы:")
print(X)
print("Вектор невязки: F=Ax*-b")
print(F)
print("Норма:")
print(D[0])
